### ISSUE TITLE
`format_args!` does not respect integer literals' type

### ISSUE THREAD

**Issue by endorpersand** [rust-lang/rust#115423]
When using the `1234i8` syntax in the Display format in formatting macros, the `i8` is ignored and treated as a `u128`.

I tried this code:

```rust
fn main() {
    println!("{}", 0x8Fi8);
}
```

I expected to see this happen: Getting an error that `0x8F` is out of the bounds for type `i8`.

Instead, this happened: This function compiles and prints `143`.

### Meta

`rustc --version --verbose`:
```
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-apple-darwin
release: 1.71.1
LLVM version: 16.0.5
```

</p>
</details>

**Comment by konnorandrews** [comment]
I went and looked through the `format_args!()` implementation and found where the issue comes from. The check at https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast_lowering/src/format.rs#L131 ignores the integer type, and it blindly calls `n.string()` on the `u128` value of the literal.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23115423.20.60format_args!.60.20does.20not.20respect.20integer.20literals'.20t.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by apiraino** [comment]
Did run a bisection on the reproducible provided (and duplicated #116631)

Regression in nightly-2023-04-22, specifically I _think_ commit 8bdcc62cb0362869f0e7b43a6ae4f96b953d3cbc

found 11 bors merge commits in the specified range
  commit[0] 2023-04-20: Auto merge of #110616 - m-ou-se:fmt-lang-items, r=jyn514
  commit[1] 2023-04-20: Auto merge of #109999 - m-ou-se:flatten-format-args, r=oli-obk
  commit[2] 2023-04-21: Auto merge of #110370 - c410-f3r:dqewdas, r=petrochenkov
  commit[3] 2023-04-21: Auto merge of #110431 - jsoref:spelling-src-etc, r=Mark-Simulacrum
  commit[4] 2023-04-21: Auto merge of #110636 - matthiaskrgr:rollup-faa33c6, r=matthiaskrgr
  commit[5] 2023-04-21: Auto merge of #96840 - cjgillot:query-feed, r=oli-obk
  commit[6] 2023-04-21: Auto merge of #109002 - michaelvanstraten:master, r=petrochenkov
  commit[7] 2023-04-21: Auto merge of #110542 - petrochenkov:qcstore4, r=cjgillot
  commit[8] 2023-04-21: Auto merge of #110569 - saethlin:mir-pass-cooperation, r=cjgillot
  commit[9] 2023-04-21: Auto merge of #110107 - cjgillot:const-prop-lint-junk, r=oli-obk
  commit[10] 2023-04-21: Auto merge of #110648 - Dylan-DPC:rollup-em3ovcq, r=Dylan-DPC

cc @m-ou-se

**Comment by m-ou-se** [comment]
Looks like it's not easy invoke the same integer-literal-range-checking code from the place where format_args!() is expanded. So the easiest solution is probably to expand to something like `{ literal; literal; fmt::Arguments::new(..) }` when literals are removed, such that the literals are still be checked later.

**Comment by m-ou-se** [comment]
Fix: https://github.com/rust-lang/rust/pull/118659
